pkg/store/proxy.go

```diff
@@ -347,16 +356,35 @@ func startStreamSeriesSet(
 	go func() {
 		defer wg.Done()
 		defer close(s.recvCh)
-
 		for {
-			r, err := s.stream.Recv()
+			frameTimeoutCtx, cancel := startFrameCtx(s.responseTimeout)
+			if cancel != nil {
+				defer cancel()
+			}
+			rCh := make(chan *recvResponse, 1)
+			var rr *recvResponse
+			go func() {
+				r, err := s.stream.Recv()
+				rCh <- &recvResponse{r: r, err: err}
```

Comment:
Consider adding `ctx.Done()` check in here.

I think the error is in the this scenario:

1. stream.Recv() blocks waiting for data
2. Client cancels request ctx.Done()
3. we close the channel (line 380)
4. stream.Recv() finishes and tries to push data into closed channel which should panic out.

Reply:
Therefore used buffered channel. Handing ctx.Done in goroutine too complicated code in my opinion.

Reply:
I don't understand what this has to do with buffered channels. Sending data to closed channel will cause panics with or without buffered channels.

Reply:
Oh, sorry, buffered channel fix another problem. In your use-case: goroutine can not send data to closed channel, because it's create for every recv. We can create channel outside of loop and try to reuse it, but it will complicated code in my opinion.

Reply:
Oh, sorry, buffered channel solve another problem)

In your use-case:
1. stream.Recv() blocks waiting for data
2. Client cancels request ctx.Done()
3. we close the channel (line 380) - only if we read from channel (in other select cases we only `return`), therefor pointer on rCh was saved in goroutine and closed(via gabarge collector) after close goroutine.
4. stream.Recv() finishes and tries to push data into not closed channel.

But I move close channel into goroutine to code will be more understandable.